engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeLengthNode.java

```diff
@@ -0,0 +1,64 @@
+package org.enso.interpreter.runtime.data.vector;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached.Exclusive;
+import com.oracle.truffle.api.dsl.NeverDefault;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.error.PanicException;
+
+public abstract class ArrayLikeLengthNode extends Node {
+  public abstract long executeLength(Object arrayLike);
+
+  @NeverDefault
+  public static ArrayLikeLengthNode create() {
+    return ArrayLikeLengthNodeGen.create();
+  }
+
+  //
+  // implementation
+  //
+
+  @Specialization
+  static long arrayLength(Array self) {
+    return self.length();
+  }
+
+  @Specialization
+  static long vectorLongLength(Vector.Long self) {
+    return self.getArraySize();
+  }
+
+  @Specialization
+  static long vectorDoubleLength(Vector.Double self) {
+    return self.getArraySize();
+  }
+
+  @Specialization
+  static long vectorLength(
+      Vector.Generic self, @Exclusive @CachedLibrary(limit = "3") InteropLibrary iop) {
+    try {
+      return self.getArraySize(iop);
+    } catch (UnsupportedMessageException ex) {
+      throw CompilerDirectives.shouldNotReachHere(ex);
+    }
+  }
+
+  @Specialization
```

Comment:
shouldn't this be guarded by `hasArrayElements` instead?

Reply:
Throwing `UnsupportedMessageException` is a fast-path operation - e.g. I think we don't need the check. But I am making this a `@Fallback` to be used when other specializations aren't applicable in e7b219dfef